### DISCUSSION TITLE
How to get best two 'alternatives' in Deepgram Self-hosted Streaming speech-to-text setup

### DISCUSSION THREAD

**Discussion by devalbham** [deepgram/community#1022] [General help] [answered]
Hi, My setup is:
- Deepgram self-hosted model
- Streaming live audio and getting transcription using Websockets.

Basically I want to get two best probabilities for transcription. For example in below reply the best on is with confidence: 0.95265996
So I also want to get next one with confidence just below it.  

Stack is React and Python FastAPI

Request ID: 151c83cd-94c1-4c0a-bc2a-26193729e887

I get sample response: Full response from Deepgram: {'type': 'Results', 'channel_index': [0, 1], 'duration': 1.79, 'start': 2.67, 'is_final': True, 'speech_final': True, 'channel': {'alternatives': [{'transcript': 'Hello. Hello. Hello. Can you hear me?', 'confidence': 0.95265996, 'words': [{'word': 'hello', 'start': 2.91, 'end': 3.23, 'confidence': 0.8148965, 'speaker': 0, 'punctuated_word': 'Hello.'}, {'word': 'hello', 'start': 3.23, 'end': 3.5500002, 'confidence': 0.952556, 'speaker': 0, 'punctuated_word': 'Hello.'}, {'word': 'hello', 'start': 3.5500002, 'end': 3.63, 'confidence': 0.83288157, 'speaker': 0, 'punctuated_word': 'Hello.'}, {'word': 'can', 'start': 3.63, 'end': 3.79, 'confidence': 0.95265996, 'speaker': 0, 'punctuated_word': 'Can'}, {'word': 'you', 'start': 3.79, 'end': 4.11, 'confidence': 0.9997596, 'speaker': 0, 'punctuated_word': 'you'}, {'word': 'hear', 'start': 4.11, 'end': 4.19, 'confidence': 0.99995744, 'speaker': 0, 'punctuated_word': 'hear'}, {'word': 'me', 'start': 4.19, 'end': 4.46, 'confidence': 0.99896395, 'speaker': 0, 'punctuated_word': 'me?'}]}]}, 'metadata': {'request_id': '151c83cd-94c1-4c0a-bc2a-26193729e887', 'model_info': {'name': '2-medical-nova', 'version': '2024-06-11.4692', 'arch': 'nova-2'}, 'model_uuid': '46658e06-4884-4582-9bc2-e92ef7baa396'}, 'from_finalize': False}

I am looking for an option to get top two transcription '**alternatives**' as response, instead of just one:

I tried: 
            # Construct the Deepgram WebSocket URL with the dynamic model
            DEEPGRAM_WS_URL = (
                f"{DG_URL}"
                f"?model={DG_MODEL}"
                "&type=streaming"
                "&diarize=true"
                "&smart_format=true"
                "&alternatives=2"
            )
            
However, doesn't look like it's a correct option and DG stop sending transcription if I use alternatives option in it. 
So how do i achieve this?

**Comment by deepgram-community** [comment]
Hey there! It looks like you haven't connected your GitHub account to your Deepgram account. You can do this at https://community.deepgram.com - being verified through this process will allow our team to help you in a much more streamlined fashion.

**Comment by deepgram-community** [comment]
Looking at our logs I'm seeing you sent a request as:

`/v1/listen?model=nova-2-medical&type=streaming&diarize=true&smart_format=true`

And I see this error:
```Error: Unexpected query parameter: type=['streaming']```
Type Streaming isn't a valid API option, just something to note. We still processed this request it seems as we tend to ignore invalid API Params.

We have a deprecated feature called `alternatives` you can try in this case. It's no longer documented as it's something we are phasing out of our API.

Here's a reference to the <https://github.com/deepgram/deepgram-js-sdk/blob/18b46406d288d076f57a909b76c590910a96ec2a/src/lib/types/SyncPrerecordedResponse.ts#L6|alternatives interface> in our JS SDK

so to make this request try to get an alternative transcript try:

`/v1/listen?model=nova-2-medical&diarize=true&smart_format=true&alternative=true`

It should still work, but like I said it's technical a deprecated feature we haven't yet removed from our API.

> ![avatar](https://avatars.slack-edge.com/2023-02-20/4855338233600_fad9174023ee3f640ce1_32.png) This message was sent by John Vajda from Deepgram, via our community automation.

**Reply by devalbham** [reply]
Thanks John, I tried with you suggested changes, however still getting back only single alternative in the response. Sample response below. Thanks.

 Full response from Deepgram: {'type': 'Results', 'channel_index': [0, 1], 'duration': 3.12, 'start': 0.0, 'is_final': True, 'speech_final': True, 'channel': {'alternatives': [{'transcript': "**Hello. Hello. I'm trying to say something. Can you hear me?**", 'confidence': 0.9996735, 'words': [{'word': 'hello', 'start': 0.64, 'end': 0.96, 'confidence': 0.9394239, 'speaker': 0, 'punctuated_word': 'Hello.'}, {'word': 'hello', 'start': 0.96, 'end': 1.28, 'confidence': 0.9990798, 'speaker': 0, 'punctuated_word': 'Hello.'}, {'word': "i'm", 'start': 1.28, 'end': 1.36, 'confidence': 0.9979826, 'speaker': 0, 'punctuated_word': "I'm"}, {'word': 'trying', 'start': 1.36, 'end': 1.68, 'confidence': 0.9997013, 'speaker': 0, 'punctuated_word': 'trying'}, {'word': 'to', 'start': 1.68, 'end': 1.8399999, 'confidence': 0.99983156, 'speaker': 0, 'punctuated_word': 'to'}, {'word': 'say', 'start': 1.8399999, 'end': 2.0, 'confidence': 0.9995654, 'speaker': 0, 'punctuated_word': 'say'}, {'word': 'something', 'start': 2.0, 'end': 2.32, 'confidence': 0.9996735, 'speaker': 0, 'punctuated_word': 'something.'}, {'word': 'can', 'start': 2.32, 'end': 2.48, 'confidence': 0.9999223, 'speaker': 0, 'punctuated_word': 'Can'}, {'word': 'you', 'start': 2.48, 'end': 2.6399999, 'confidence': 0.9996778, 'speaker': 0, 'punctuated_word': 'you'}, {'word': 'hear', 'start': 2.6399999, 'end': 2.8, 'confidence': 0.9999641, 'speaker': 0, 'punctuated_word': 'hear'}, {'word': 'me', 'start': 2.8, 'end': 3.12, 'confidence': 0.99665624, 'speaker': 0, 'punctuated_word': 'me?'}]}]}, 'metadata': {'request_id': 'e5803fda-3781-4d6f-b897-80256c07d366', 'model_info': {'name': '2-medical-nova', 'version': '2024-06-11.4692', 'arch': 'nova-2'}, 'model_uuid': '46658e06-4884-4582-9bc2-e92ef7baa396'}, 'from_finalize': False}

**Comment by deepgram-community** [comment, marked as answer]
I did some testing,  It seems Alternatives does work for our older models specifically `nova` , `base` , `enhanced`  NOT `nova-2` . Also instead of `true` you'd need to pass in an integer value for the number of alternatives you want back.

Here's an example cURL that worked for me:

```
curl --location '<https://api.deepgram.com/v1/listen?alternatives=2&model=nova>' \
--header 'Content-Type: audio/wave' \
--header 'Authorization:  Your API KEY' \
--data-binary '@/YOUR FILE'
```

> ![avatar](https://avatars.slack-edge.com/2023-02-20/4855338233600_fad9174023ee3f640ce1_32.png) This message was sent by John Vajda from Deepgram, via our community automation.
